registry/federated-namespaces.md

```diff
@@ -0,0 +1,155 @@
+# Proposal: Federating namespace restrictions
+
+This proposes minimizing the constraints the Docker client places on image names, deferring instead to the Registry at push/pull time to push back with a suitable error message around character restrictions / form.
+
+
+## Preface: Current State
+
+Today, the client is overly aware of the character restrictions and form of repository names.  These restrictions have been adjusted throughout the course of v1, and have changed non-trivially in the v1 => v2 switch.
+
+The restrictions enforced in the client today are very v1-centric, which makes some of the relaxations in v2 unusable in their current form.
+
+### v1
+
+In the world of Docker Registry v1, names have the form:
+```
+   registry/namespace/image:tag
+```
+
+`tag` can often be omitted, and depending on context can mean `latest` or all tags.
+
+`registry` can be omitted, and defaults to DockerHub (`index.docker.io`).
+
+`namespace` can be omitted, defaulting to `library`.
+
+`image` is required.
+
+
+This image path is then processed as follows:
+   1. If there is a single part, continue processing as `index.docker.io/library/single-part`
+   1. If there are two parts, and the first part is `localhost` or contains a `:` or `.` character, continue processing as `first-part/library/second-part`
+   1. If there are more than three parts, fail.
+
+The result of the above should result in exactly three parts when split on `/`:
+   1. `registry`: The hostname of the registry with which to interact.
+   1. `namespace`: The namespace (on DockerHub the user or organization) under which the image exists.
+   1. `image:tag`: This is further split on `:` into at most 2 pieces, with `tag` defaulting to `latest`.
+
+These components are validated against the regular expressions (see [validateRemoteName](https://sourcegraph.com/github.com/docker/docker@624de8a9cd5652d0d685596d490997f7f2e83536/.tree/registry/config.go#startline=205&endline=205), see [ValidateTagName](https://sourcegraph.com/github.com/docker/docker@624de8a9cd5652d0d685596d490997f7f2e83536/.tree/graph/tags.go#startline=354&endline=354)):
+   1. `namespace` must match `^([a-z0-9-_]*)$`
+   1. `image` must match `^([a-z0-9-_.]+)$`
+   1. `tag` must match `^[\w][\w.-]{0,127}$`
+
+### v2
+
+In the world of Docker Registry v2, names have a similar, but slightly more flexible form best described [here](https://sourcegraph.com/github.com/docker/distribution@5556cd1ba11aea6803f3c51a16bb96813ba72800/.tree/registry/api/v2/names.go#selected=56):
+```
+// Effectively, the name should comply with the following grammar:
+//
+// 	alpha-numeric := /[a-z0-9]+/
+//	separator := /[._-]/
+//	component := alpha-numeric [separator alpha-numeric]*
+//	namespace := component ['/' component]*
+```
+Where `namespace` is used in the context: `registry/namespace:tag`, where as before `registry` and `tag` are optional.
```

Comment:
V2 does not prescribed this definition of namespace. This description is more of the conventions used on the hub. The number of components is not restricted and the meaning of each component is use case specific.